File: src/Linkedlist/CircularLinkedList.java
package Linkedlist;

public class CircularLinkedList {
    private Node head;
    private Node tail;
    public CircularLinkedList(){
        this.head = null;
        this.tail = null;
    }
    public void insert(int value){
        Node node = new Node(value);
        if(head==null){
            head = node;
            tail = node;
            return;
        }
        node.next = head;
        tail.next = node;
        tail = node;
    }
    public void display(){
        Node temp = head;
        if(temp!=null){
            do{
                System.out.print(temp.value+"->");
                if (temp.next != null) {
                    temp = temp.next;
                }
            }while(temp!=head);
            System.out.println("HEAD");
        }
    }
    public void delete(int value){
        if(head==null){
            return;
        }
        if(head.value==value){
            head = head.next;
            tail.next = head;
            return;
        }
        Node node = head;
        do{
            Node nextNode = node.next;
            if(nextNode.value==value){
                node.next = nextNode.next;
                if(tail==nextNode){
                    tail=node;
                }
                return;
            }
            node = node.next;
        }while(node!=head);
    }
    private class Node{
        int value;
        Node next;
        public Node(int value){
            this.value = value;
        }
    }
}

File: pythonFiles/LinkedList/CircularLinkedList.py
class CircularLinkedList:
    class Node:
        def __init__(self,data):
            self._data = data
            self._next = None
    def __init__(self):
        self._head = None
        self._tail = None
        self._size = 0
    def insert(self,value):
        node = self.Node(value)
        if self._head==None:
            self._head = node
        else:
            self._tail._next = node
        self._tail = node
        node._next = self._head
        self._size += 1
        return self._size
    def display(self):
        if(self._head!=None):
            temp = self._head
            print(str(temp._data)+"->",end="")
            temp = temp._next
            while(temp!=self._head):
                print(str(temp._data)+"->",end="")
                temp = temp._next
            print("HEAD")
        else:
            print("None")
    def delete(self,value):
        node = self.find(value)
        if(node==None):
            return self._size
        prev = self._head
        while(prev._next!=node):
            prev = prev._next
        if(prev==node):
            self._head = None
            self._tail = None
        else:
            prev._next = node._next
            if(node==self._head):
                self._head = prev._next
            elif(node==self._tail):
                self._tail = prev
        self._size += 1
        return self._size
    def find(self,value):
        node = self._head
        if(node!=None):
            if(node._data == value):
                return node
            node = node._next
            while(node!=self._head):
                if(node._data == value):
                    return node
                node = node._next
        return None   
cll = CircularLinkedList()
cll.display()
cll.insert(1)
cll.display()
cll.delete(1)
cll.display()
cll.insert(2)
cll.display()
cll.insert(3)
cll.delete(3)
cll.display()
cll.insert(3)
cll.insert(4)
cll.display()
cll.delete(3)
cll.display()